File: 2961_도영이가만든맛있는음식.py
import sys
n = int(sys.stdin.readline())

bitter = []
sour = []

for __ in range(n):
    s,b = map(int, sys.stdin.readline().split())
    bitter.append(b)
    sour.append(s)
used = [False] * n

answer = 1<<32-1
for bitmask in range(1,(1<<n)):
    b = 0
    s = 1
    for i in range(n):
        if (1<<i)&bitmask:
            b += bitter[i]
            s *= sour[i]
    if answer > abs(b-s):
        answer = abs(b-s)
print(answer)

File: Java/2961_도영이가만든맛있는음식.java

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

public class B2961 {

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		int n = Integer.parseInt(br.readLine());
		int[] bitter = new int[n];
		int[] sour = new int[n];
		for(int i=0;i<n;i++) {
			StringTokenizer token = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(token.nextToken());
			int b = Integer.parseInt(token.nextToken());
			bitter[i] = b;
			sour[i] = s;
		}
		int answer = (1<<31)-1;
		for(int bitmask=1; bitmask!=(1<<n); bitmask++) {
			int bSum = 0;
			int sSum = 1;
			for(int i=0; i<n; i++) {
				if(((1<<i)&bitmask)!=0) {
					bSum += bitter[i];
					sSum *= sour[i];
				}
			}
			if(Math.abs(bSum-sSum) < answer)
				answer = Math.abs(bSum-sSum);
		}

		bw.write(Integer.toString(answer));
		bw.flush();
	}

}
